src/System.Net.Http/src/System/Net/Http/Managed/AuthenticationHelper.Digest.cs

```diff
@@ -0,0 +1,374 @@
+            {
+                algorithm = Sha256;
+            }
+
+            string realm = digestResponse.Parameters.ContainsKey(Realm) ? digestResponse.Parameters[Realm] : string.Empty;
+
+            // Add username
+            if (digestResponse.Parameters.ContainsKey(UserHash) && digestResponse.Parameters[UserHash] == "true")
+            {
+                sb.AppendKeyValue(Username, ComputeHash(digestResponse.Parameters[Username] + Colon + realm, algorithm));
+            }
+            else
+            {
+                sb.AppendKeyValue(Username, credential.UserName);
+            }
+
+            // Add realm
+            if (realm != string.Empty)
+                sb.AppendKeyValue(Realm, realm);
+
+            // If nonce is same as previous request, update nonce count.
+            if (digestResponse.Parameters[Nonce] == digestResponse.Nonce)
+            {
+                digestResponse.NonceCount++;
+            }
+
+            // Add nonce
+            sb.AppendKeyValue(Nonce, digestResponse.Parameters[Nonce]);
+            digestResponse.Nonce = digestResponse.Parameters[Nonce];
+
+            // Add uri
+            sb.AppendKeyValue(Uri, request.RequestUri.PathAndQuery);
+
+            // Set qop, default is auth
+            string qop = Auth;
+            if (digestResponse.Parameters.ContainsKey(Qop))
+            {
+                // Check if auth-int present in qop string
+                int index1 = digestResponse.Parameters[Qop].IndexOf(AuthInt);
+                if (index1 != -1)
+                {
+                    // Get index of auth if present in qop string
+                    int index2 = digestResponse.Parameters[Qop].IndexOf(Auth);
+
+                    // If index2 < index1, auth option is available
+                    // If index2 == index1, check if auth option available later in string after auth-int.
+                    if (index2 == index1)
+                    {
+                        index2 = digestResponse.Parameters[Qop].IndexOf(Auth, index1 + AuthInt.Length);
+                        if (index2 == -1)
+                        {
+                            qop = AuthInt;
+                        }
+                    }
+                }
+            }
+
+            // Set cnonce
+            string cnonce = GetRandomAlphaNumericString();
+
+            // Calculate response
+            string a1 = credential.UserName + Colon + realm + Colon + credential.Password;
+            if (digestResponse.Parameters[Algorithm].Contains(Sha256Sess))
+            {
+                a1 = ComputeHash(a1, algorithm) + Colon + digestResponse.Parameters[Nonce] + Colon + cnonce;
+            }
+
+            string a2 = request.Method.Method + Colon + request.RequestUri.PathAndQuery;
+            if (qop == AuthInt)
+            {
+                a2 = a2 + Colon + ComputeHash(request.Content.ReadAsStringAsync().Result, algorithm);
```

Comment:
@mconnew  I didn't understand this part correctly from the RFC, the RFC states use entity-body, which from the definition looked like they are referring to the content of the request. So what should i do here? In cases of null content, i dunno what to use, the RFC doesn't say anything about that, should i use empty string? Not sure how to handle large content.